File: Python/DesignPattern/CreationalDesignPattern/CompositionDesignPattern.py
from abc import ABC, abstractmethod
from typing import List


class IFileSystemNode(ABC):
    @abstractmethod
    def get_size(self) -> int: pass

    @abstractmethod
    def get_name(self) -> str: pass

    @abstractmethod
    def tree(self): pass


class File(IFileSystemNode):
    def __init__(self, name: str, content: str):
        self._name = name
        self._content = content

    def get_size(self) -> int:
        return len(self._content)

    def get_name(self) -> str:
        return self._name

    def tree(self):
        print(f"File {self._name}")


class Folder(IFileSystemNode):
    def __init__(self, name: str):
        self._name = name
        self._children: List[IFileSystemNode] = []

    def touch(self, name: str, content: str):
        self._children.append(File(name, content))

    def mkdir(self, name: str):
        self._children.append(Folder(name))

    def get_size(self) -> int:
        return sum(child.get_size() for child in self._children)

    def get_name(self) -> str:
        return self._name

    def tree(self):
        print(f"Folder:-> {self._name}")
        for child in self._children:
            child.tree()


if __name__ == "__main__":
    root = Folder("root")
    root.touch("System.txt", "dasdsadasd")
    root.mkdir("Doc")

File: C++/DesignPattern/CreationalDesignPattern/CompositionDesignPattern.cpp
#include <iostream>
#include <string>
#include <vector>
#include <memory>
using namespace std;

struct IFileSystemNode {
    virtual int getSize() const = 0;
    virtual string getName() const = 0;
    virtual void tree() const = 0;
    virtual ~IFileSystemNode() = default;
};

class File : public IFileSystemNode {
    string name, content;
public:
    File(string name, string content) : name(move(name)), content(move(content)) {}

    int getSize() const override { return (int)content.size(); }
    string getName() const override { return name; }
    void tree() const override { cout << "File " << name << "\n"; }
};

class Folder : public IFileSystemNode {
    string name;
    vector<unique_ptr<IFileSystemNode>> children;
public:
    explicit Folder(string name) : name(move(name)) {}

    void touch(const string& n, const string& content) {
        children.push_back(make_unique<File>(n, content));
    }

    void mkdir(const string& n) {
        children.push_back(make_unique<Folder>(n));
    }

    int getSize() const override {
        int size = 0;
        for (const auto& child : children) size += child->getSize();
        return size;
    }

    string getName() const override { return name; }

    void tree() const override {
        cout << "Folder:-> " << name << "\n";
        for (const auto& child : children) child->tree();
    }
};

int main() {
    Folder root("root");
    root.touch("System.txt", "dasdsadasd");
    root.mkdir("Doc");
    return 0;
}
